content/blog/2025-04-09-introducing-bloom-filters.md

```diff
@@ -0,0 +1,154 @@
+date= 2025-04-09 01:01:01
+authors= [ "karthiksubbarao"]
++++
+
+The Valkey project is introducing Bloom Filters as a new data type via [valkey-bloom](https://github.com/valkey-io/valkey-bloom/), an official Valkey Module which is compatible with Valkey versions >= 8.0. Bloom filters provide efficient, large-scale membership testing, improving performance and offering significant memory savings for high-volume applications.
+
+As an example, to handle advertisement deduplication workloads and answer the question, "Has this customer seen this ad before?", previously, Valkey developers could use the SET data type.
+This is done by adding the customer IDs (of those who viewed an ad) into a `SET` object representing a particular advertisement. However, since every item in the set is allocated, the problem with this approach is high memory usage.
+This article demonstrates how using the bloom filter data type from valkey-bloom achieves >93% memory savings for the same workload, while exploring its implementation, technical details, and practical recommendations.
+
+## Introduction
+
+Bloom filters are a space efficient probabilistic data structure that allows adding elements and checking whether elements exist. False positives are possible where a filter incorrectly indicates that an element exists, even though it was not added.
+However, Bloom Filters guarantee that false negatives do not occur, meaning it can never be the case that an element was added successfully and a filter reports it as not existing.
+Bloom Filters were first introduced in a paper from 1970 by Burton H. Bloom.
+
+![Bloom Filter Bit Vector](/assets/media/pictures/bloomfilter_bitvector.png)
+*Image taken from [source](https://en.wikipedia.org/wiki/Bloom_filter#/media/File:Bloom_filter.svg)*
+
+When adding an item to a bloom filter, K hash functions compute K corresponding bits from the bit vector which are set to 1.
+Checking existence involves the same hash functions - if any bit is 0, the item is definitely absent; if all bits are 1, the item likely exists (with a defined false positive probability).
+This bit-based approach, rather than full item allocation, makes bloom filters very space efficient with the trade off being potential false positives.
+
+Valkey-Bloom (BSD-3 licensed) is an official Valkey module that introduces bloom filters as a new data type to Valkey, providing both scalable and non-scalable variants.
+It is API compatible with the bloom filter command syntax of the official Valkey client libraries including valkey-py, valkey-java, valkey-go (as well as the equivalent Redis libraries).
+The module was developed by the following authors: Karthik Subbarao ([KarthikSubbarao](https://github.com/KarthikSubbarao)), Cameron Zack ([zackcam](https://github.com/zackcam)), Vanessa Tang ([YueTang-Vanessa](https://github.com/YueTang-Vanessa)), Nihal Mehta ([nnmehta](https://github.com/nnmehta)), and wuranxx ([wuranxx](https://github.com/wuranxx)).
```

Comment:
I would move the callout to the community to the bottom. I think it's important to include, but it breaks the flow for readers.